src/imperative/imperative.cc

```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
+  static const Op* copy_op = Op::Get("_copy");
+
+  Graph graph;
+  graph.outputs = CreateForwardGraph(outputs);
+
+
+  // Prepare head gradient nodes
+  std::vector<NodeEntry> ograd_entries = CreateHeadGradientNodes(outputs, ograds);
+
+  // Get variable nodes
+  GradientVariableNodes gvars = CreateGradientVariableNodes(variables, graph.outputs);
+
+  // Run backward on the graph
+  Graph gradient_graph = pass::MXGradient(
+      graph, graph.outputs, gvars.variable_nodes, ograd_entries,
       exec::AggregateGradient, nullptr, nullptr,
       zero_ops, "_copy");
-  CHECK_EQ(g_graph.outputs.size(), xs.size());
-  for (const auto& e : g_graph.outputs) {
-    if (e.node->op() == nullptr) {
+
+  CHECK_EQ(gradient_graph.outputs.size(), gvars.variable_nodes.size());
+  std::vector<nnvm::NodeEntry> forward_outputs = graph.outputs;
+  const size_t num_forward_outputs = graph.outputs.size();
+
+  // TODO(larroy): move inside pass::MXGradient
+  for (const auto& backward_node : gradient_graph.outputs) {
+    if (backward_node.node->is_variable()) {
       auto node = Node::Create();
       node->attrs.op = copy_op;
-      node->inputs.push_back(e);
+      node->inputs.push_back(backward_node);
       graph.outputs.emplace_back(std::move(node));
     } else {
-      graph.outputs.push_back(e);
+      graph.outputs.push_back(backward_node);
     }
   }
-  const auto& idx = graph.indexed_graph();
+
+  const auto& indexed_graph = graph.indexed_graph();
   // get number of nodes used in forward pass
   size_t num_forward_nodes = 0;
   size_t num_forward_entries = 0;
   for (size_t i = 0; i < num_forward_outputs; ++i) {
     num_forward_nodes = std::max(
-        num_forward_nodes, static_cast<size_t>(idx.outputs()[i].node_id + 1));
+        num_forward_nodes, static_cast<size_t>(indexed_graph.outputs()[i].node_id + 1));
     num_forward_entries = std::max(
-        num_forward_entries, static_cast<size_t>(idx.entry_id(idx.outputs()[i])) + 1);
+        num_forward_entries, static_cast<size_t>(indexed_graph.entry_id(
+            indexed_graph.outputs()[i])) + 1);
   }
 
   // Allocate buffer
-  std::vector<NDArray> buff(idx.num_node_entries());
+  std::vector<NDArray> buff(indexed_graph.num_node_entries());
   std::vector<uint32_t> ref_count(buff.size(), 0);
   std::vector<OpStatePtr> states;
   std::vector<NDArray*> arrays;
   arrays.reserve(buff.size());
-  for (auto& buffered_array : buff) {
+  for (auto& buffered_array : buff)
     arrays.push_back(&buffered_array);
-  }
+
   if (create_graph) {
     states.resize(num_forward_nodes);
-    nnvm::DFSVisit(sym.outputs, [&](const nnvm::NodePtr& n) {
+    nnvm::DFSVisit(forward_outputs, [&](const nnvm::NodePtr& n) {
       AGInfo& info = AGInfo::Get(n);
-      states[idx.node_id(n.get())] = info.state;
+      states.at(indexed_graph.node_id(n.get())) = info.state;
       for (uint32_t i = 0; i < info.outputs.size(); ++i) {
```

Comment:
change i to `size_t` to be consistent in your refactoring?

Reply:
Ok, but it had a reason for that.